aws-xray-agent/src/main/java/com/amazonaws/xray/agent/runtime/handlers/XRayHandler.java

```diff
@@ -129,33 +129,32 @@ protected void endSubsegment() {
      * all the URL, method, host, origin, and service name information.
-     * @param transactionState The current state of the X-Ray transaction.
-     * @param traceHeader An optional trace header that's passed in usually as a result of an upstream call.
+     * @param transactionState The current state of the X-Ray transaction. Includes the trace header from an upstream
+     *                         call if applicable.
      * @return True if we should sample, false otherwise.
      */
-    protected boolean getSamplingDecision(XRayTransactionState transactionState, Optional<TraceHeader> traceHeader) {
-        String method = transactionState.getMethod();
-        String serviceName = XRayTransactionState.getServiceName();
-        String host = transactionState.getHost();
-        String uri = transactionState.getURL();
-
-        String traceHeaderString = Optional.ofNullable(transactionState.getTraceHeader()).map(Object::toString).orElse(null);
+    protected boolean getSamplingDecision(XRayTransactionState transactionState) {
         // If the trace header string is null, then this is the origin call.
-        if (traceHeaderString != null) {
-            traceHeader = Optional.of(TraceHeader.fromString(traceHeaderString));
-        }
-
-        TraceHeader.SampleDecision sampleDecision = traceHeader.map(TraceHeader::getSampled).orElse(null);
-        if (sampleDecision != null) {
-            log.debug(String.format("Received sampling decision from trace header: %s", sampleDecision.toString()));
-            return sampleDecision.equals(TraceHeader.SampleDecision.SAMPLED);
+        TraceHeader traceHeader = TraceHeader.fromString(transactionState.getTraceHeader());
+
+        TraceHeader.SampleDecision sampleDecision = traceHeader.getSampled();
+        if (TraceHeader.SampleDecision.SAMPLED.equals(sampleDecision)) {
+            log.debug("Received SAMPLED decision from upstream X-Ray trace header");
+            return true;
+        } else if (TraceHeader.SampleDecision.NOT_SAMPLED.equals(sampleDecision)) {
+            log.debug("Received NOT SAMPLED decision from upstream X-Ray trace header");
+            return false;
         }
 
         // No sampling decision made on the upstream. So use the in-house rules.
-        SamplingRequest samplingRequest = new SamplingRequest(serviceName, host, uri, method,
+        SamplingRequest samplingRequest = new SamplingRequest(
+                XRayTransactionState.getServiceName(),
+                transactionState.getHost(),
+                transactionState.getURL(),
+                transactionState.getMethod(),
                 transactionState.getServiceType());
         SamplingResponse samplingResponse = AWSXRay.getGlobalRecorder().getSamplingStrategy().shouldTrace(samplingRequest);
-        return samplingResponse.isSampled();
+        return samplingResponse != null && samplingResponse.isSampled();
```

Comment:
Which situations the samplingResponse is null?

Reply:
I suppose it only really could be if a customer created a custom sampling strategy, but that's not possible to plug into the agent. I'll simplify this.